### DISCUSSION TITLE
Counterparty Functions Scheduled to a Specific Block

### DISCUSSION THREAD

**Discussion by davestaxcp** [CounterpartyXCP/CIPs#117] [CIPs]
I have noticed users looking to use Counterparty functions at a specific Bitcoin block.

Is this possible to do on Counterparty using this: https://en.bitcoin.it/wiki/NLockTime ?

The use cases done in the past were auctions for specific XCP/TOKEN pairs and at a "specific block" the auction ended

Another use case that came up was for "scheduled burns" to happen at a specific time (specific block)

Could this be possible in the future for use on Counterparty?

**Comment by Jpja** [comment]
I wonder the same. 

In my understanding, which may be wrong, the answer is NO. Bitcoin scripting does not work for Counterparty. The reason being Bitcoin scripts are utxo based. Counterparty is address based. If you time lock a specific utxo, you could always send a token from that address by spending from another utxo.

The exception is atomic swaps. Bitcoin scripting should work here. The reason being that both parties agree on a specific tx, so the underlying utxo is predetermined. 

@keyuno @loon3 care to confirm/correct?

**Reply by keyuno** [reply]
When an address is generated using a particular locking script, Pay-to-Script-Hash (P2SH) address, any transaction attempting to spend from that address must include the matching script in the input.

If you send Counterparty tokens to a P2SH address associated with a specific locking script, you must include the corresponding script in the spending transaction's input in order to spend those tokens. 

This requirement ensures that the conditions specified in the locking script are satisfied before the tokens or bitcoin can be spent, regardless of whether you're dealing with Counterparty tokens or Bitcoin itself.

**Reply by Jpja** [reply]
So the lock script is required to spend ANY utxo from that address? Ie the script puts a time lock on the address, not on a specific utxo?

**Reply by keyuno** [reply]
When funds are sent to a P2SH address, they are effectively 'locked up' according to the conditions specified in the script used to generate the address. To spend any of those funds, you must provide the matching script (also known as the redeem script) that satisfies the conditions outlined in the original script. It's not tied to a specific UTXO but rather to the address as a whole.

**Reply by Jpja** [reply]
Thanks for clarifying. Does this imply that all the clever stuff you can do with bitcoin, you can do with counterparty tokens too? Including lightning network?

**Reply by keyuno** [reply]
I haven't delved deeply into lightning network, but I'm aware of past experiments that explored payment channels. 

https://github.com/CounterpartyXCP/picopayments-hub

**Comment by jotapea** [comment]
Any function which only affect assets (the virtual token universe), should be possible to add to Counterparty. Bitcoin UTXO integration is separate, and explained above...
